File: python/146_LRU_Cache.py
# Design and implement a data structure for Least Recently Used (LRU) cache.
# It should support the following operations: get and set.

# get(key) - Get the value (will always be positive) of the key
# if the key exists in the cache, otherwise return -1.
# set(key, value) - Set or insert the value if the key is not already present. 
# When the cache reached its capacity, it should invalidate the least recently
# used item before inserting a new item.

# Runtime: 196/460 ms

# Implemented by Double Linked List and Hashtable
class ListNode(object):
    def __init__(self, val):
        self.val = val
        self.next = self
        self.prev = self
    def append_before(self, node):
        tmp = self.prev
        self.prev = node
        node.prev = tmp
        tmp.next = node
        node.next = self
    def append_after(self, node):
        tmp = self.next
        self.next = node
        node.next = tmp
        tmp.prev = node
        node.prev = self
    def delete_before(self):
        if self.prev == self:
            raise ValueError("only one node")
        tmp = self.prev.prev
        tmp.next = self
        self.prev = tmp
    def delete_after(self):
        if self.next == self:
            raise ValueError("only one node")
        tmp = self.next.next
        tmp.prev = self
        self.next = tmp

class LRUCache(object):
    def __init__(self, capacity):
        """
        :type capacity: int
        """
        self.capacity = capacity
        self.c_size = 0
        self.history = ListNode('dummy')
        self.hashtable = {}
        self._data = {}   

    def _update_history(self, key):
        if self.hashtable.get(key, None) is None:
            # this key is not in the hashtable
            # create the node and append to history
            this_node = ListNode(key)
            self.hashtable[key] = this_node
            self.history.append_before(this_node)
        else:
            # this key is already in the hashtable
            # delete the old node from history and then append it
            this_node = self.hashtable[key]
            this_node.next.delete_before()
            self.history.append_before(this_node)

    def get(self, key):
        """
        :rtype: int
        """
        if self._data.get(key, None) is None:
            return -1
        else:
            self._update_history(key)
        return self._data.get(key)
        

    def set(self, key, value):
        """
        :type key: int
        :type value: int
        :rtype: nothing 
        """
        if self._data.get(key,-1) == -1:
            # the key doesn't exist, we need to add new data
            if self.c_size == self.capacity:
                # the cache is full, need to delete one node
                tobeDelete = self.history.next
                self.history.delete_after()
                self.hashtable.pop(tobeDelete.val)
                self._data.pop(tobeDelete.val)
            else:
                # update the size counter
                self.c_size += 1
            # add new data
            self._update_history(key)
            self._data[key] = value
        else:
            # update old value
            self._update_history(key)
            self._data[key] = value




# Implement by List
class LRUCache_On_List_sol(object):

    def __init__(self, capacity):
        """
        :type capacity: int
        """
        self.capacity = capacity
        self.c_size = 0
        self.history = []
        self._data = {}
        

    def get(self, key):
        """
        :rtype: int
        """
        if self._data.get(key, None) is None:
            return -1
        else:
            self.history.remove(key)
            self.history.append(key)
        return self._data.get(key)
        

    def set(self, key, value):
        """
        :type key: int
        :type value: int
        :rtype: nothing
        """
        if self._data.get(key,-1) == -1:
            if self.c_size == self.capacity:
                # first delete the LRU data
                delkey = self.history.pop(0)
                self._data.pop(delkey)
                # then add new data
                self._data[key] = value
                self.history.append(key)

            else:
                # add new data
                self.c_size += 1
                self.history.append(key)
                self._data[key] = value
        else:
            # update old value
            self._data[key] = value
            self.history.remove(key)
            self.history.append(key)

File: java/146_LRU_Cache.java
// Design and implement a data structure for Least Recently Used (LRU) cache. It should support the following operations: get and put.

// get(key) - Get the value (will always be positive) of the key if the key exists in the cache, otherwise return -1.
// put(key, value) - Set or insert the value if the key is not already present. When the cache reached its capacity, it should invalidate the least recently used item before inserting a new item.

// Follow up:
// Could you do both operations in O(1) time complexity?

// Example:

// LRUCache cache = new LRUCache( 2 /* capacity */ );

// cache.put(1, 1);
// cache.put(2, 2);
// cache.get(1);       // returns 1
// cache.put(3, 3);    // evicts key 2
// cache.get(2);       // returns -1 (not found)
// cache.put(4, 4);    // evicts key 1
// cache.get(1);       // returns -1 (not found)
// cache.get(3);       // returns 3
// cache.get(4);       // returns 4


// Runtime: 138 ms


public class LRUCache {
    
    private ListNode head; // point to the dummy node, whose next node is LRU
    private ListNode tail; // point to the MRU node
    private Map<Integer, ListNode> cache;
    private int capacity;

    public LRUCache(int capacity) {
        this.capacity = capacity;
        this.cache = new HashMap<>();
        this.head = new ListNode(-1, -1);
        this.tail = this.head;
    }
    
    public int get(int key) {
        int res = -1;
        if (cache.containsKey(key)) {
            ListNode thisNode = cache.get(key);
            res = thisNode.val;
            updateNode(thisNode.key, thisNode.val);
        }
        return res;
    }
    
    public void put(int key, int val) {
        if (cache.containsKey(key)) {
            updateNode(key, val);
        } else if (cache.size() < capacity) {
            addNode(key, val);
        } else {
            removeNode();
            addNode(key, val);
        }
    }
    
    private void updateNode(int key, int val) {
        ListNode thisNode = cache.get(key);
        thisNode.val = val;
        // If current node is already MRU, just return
        if (thisNode.next == null) return;
        // Remove thisNode from the list
        ListNode nextNode = thisNode.next;
        thisNode.key = nextNode.key;
        thisNode.val = nextNode.val;
        thisNode.next = nextNode.next;
        cache.put(thisNode.key, thisNode);
        if (nextNode == this.tail) {
            this.tail = thisNode;
        }
        // Move nextNode to the tail of the list
        nextNode.key = key;
        nextNode.val = val;
        nextNode.next = null;
        this.tail.next = nextNode;
        this.tail = nextNode;
        cache.put(key, nextNode);
    }
    
    private void removeNode() {
        ListNode rmNode = head.next;
        head.next = rmNode.next;
        cache.remove(rmNode.key);
    }
    
    private void addNode(int key, int val) {
        ListNode newNode = new ListNode(key, val);
        this.tail.next = newNode;
        this.tail = newNode;
        cache.put(key, newNode);
    }
    
}

class ListNode {
    public ListNode next;
    public int key;
    public int val;
    public ListNode(int key, int val) {
        this.key = key;
        this.val = val;
    }
}

/**
 * Your LRUCache object will be instantiated and called as such:
 * LRUCache obj = new LRUCache(capacity);
 * int param_1 = obj.get(key);
 * obj.put(key,value);
 */